### ISSUE TITLE
[Workshop] Technical Interview Preparation

### ISSUE THREAD

**Issue by illicitonion** [CodeYourFuture/syllabus#175]
# Workshop Request

## What is the title of your workshop

Technical Interview Preparation

## What are the key topics that will be covered in the workshop?

* Typical structures of technical interviews
* How to approach problems
* Important things to remember/focus on (e.g. asking clarifying questions, explaining what you're thinking, explaining what you're stuck on, clear naming, ...)
* Handling changing constraints, or looking for improvements to your code (a lot of technical interviews take a form "Solve a problem", "Now here's an extra constraint, how does your code handle that?", which is a skill we haven't really given trainees much experience with)
* Worked example question(s), maybe actually asking some in small groups

As a stretch goal, we _may_ want to try to touch on some hand-waving efficiency (e.g. showing two solutions to a problem, one of which memoises to avoid an array traversal inside a loop, or something), and discussing how one solution is "faster" or "uses less memory" than another, but definitely not going as far as trying to formalise this (i.e. no big-O).

## What knowledge does the student need before starting?

Given a description of a set of inputs/behaviours/outputs, to be able to write a function to implement the behaviour. Example problems in terms of scope/complexity are:
* Find the longest word in an array of strings, where each string may contain arbitrarily many words
* Compute the longest prefix common to {all,any two} strings in an array
* Parse a string of Roman numerals to decimal
* Write an object which has a method to tell it the winner and loser of a game, and returns a ranking of all known players

In particular, this requires knowledge of:
* All of JS1 (functions, strings, arrays, general datatypes, comparisons, control flow)
* Keeping and mutating state in objects

## What are the topics that will **not** be covered in the workshop?

* Any coding techniques. This workshop covers approaches, and context, but assumes pre-existing knowledge of all relevant techniques in any programming language (and will be delivered assuming JavaScript is the language known).
* "System design" style interviews - though some of the techniques cross-apply, open-ended systems design requires a lot more context than solving relatively narrowly defined coding problems.

## Any other notes?

Some discussion happened on https://github.com/CodeYourFuture/syllabus/issues/158 around this, from which this issue spun out. In particular, we should work out how much of methods/object state we want to teach in JS2, and if that amount is little/none, we should work out where we _should_ teach this content.

We should probably perform a survey of what knowledge is required to successfully interview at different companies, to make sure we're covering roughly the range we're targeting.

## Completion

When completed the lesson plan, slides and exercises should be added to the Workshop section of the Syllabus

**Comment by SallyMcGrath** [comment]
This may actually be quite a pressing issue. What would you need to start this?

**Comment by ChrisOwen101** [comment]
> A survey of employers to see what kind of technical interviews they ask, and what their expectations are.

This is something that Barny (ex-Manchester City Coordinator, currently moving to Head of Employability) will be able to help us with

**Comment by SallyMcGrath** [comment]
If you can start with some assumptions, and list them at the top so it's really clear, I think we can work on getting you some real data at the same time.

After React is exactly where I was thinking too.

@Barny-6

**Comment by Barny-6** [comment]
Agreed this is a really important area. Hoping to get some hiring managers as advisors to give us some guidance here. As Sally says if you can make a start with some assumptions once I've moved roles I'll get stuck in with collecting some real data!
